File: code/java/medium/0973-kClosest.java
class Solution {
    class Point implements Comparable<Point> {
        int x;
        int y;
        int dist;

        Point(int x, int y) {
            this.x = x;
            this.y = y;
            dist = x * x + y * y;
        }

        @Override
        public int compareTo(Point p2) {
            return this.dist - p2.dist;
        }
    }

    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<Point> pq = new PriorityQueue<>();

        for (int[] point : points) {
            pq.add(new Point(point[0], point[1]));
        }

        ArrayList<ArrayList<Integer>> res = new ArrayList<>();
        for (int i = 0; i < k; i++) {
            Point curr = pq.remove();
            ArrayList<Integer> pair = new ArrayList<>();
            pair.add(curr.x);
            pair.add(curr.y);
            res.add(pair);
        }

        int[][] result = new int[k][2];
        for (int i = 0; i < k; i++) {
            result[i][0] = res.get(i).get(0);
            result[i][1] = res.get(i).get(1);
        }
        return result;
    }
}

File: code/cpp/medium/0973-kClosest.cpp
class Solution {
    public:
        vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
            auto comp = [](const pair<int, int>& a, const pair<int, int>& b) {
                return a.first * a.first + a.second * a.second >
                       b.first * b.first + b.second * b.second;
            };
    
            priority_queue<pair<int, int>, vector<pair<int, int>>, decltype(comp)>
                pq(comp);
            for (auto point : points) {
                pq.push({point[0], point[1]});
            }
    
            vector<vector<int>> res;
            for (int i = 0; i < k; i++) {
                auto curr = pq.top();
                pq.pop();
                res.push_back({curr.first, curr.second});
            }
    
            return res;
        }
    };